Services.Test/CustomDeviceModelsTest.cs

```diff
@@ -0,0 +1,342 @@
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItThrowsExternalDependencyExceptionWhenGetAllDeviceModelFailed()
+        {
+            // Arrange
+            this.storage
+                .Setup(x => x.GetAllAsync(It.IsAny<string>()))
+                .ThrowsAsync(new Exception());
+
+            // Act & Assert
+            Assert.ThrowsAsync<ExternalDependencyException>(() => this.target.DeleteAsync(It.IsAny<string>()));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItThrowsExceptionWhenGetListOfDeviceModelsDeserializeFailed()
+        {
+            // Arrange
+            this.SetupAListOfInvalidDeviceModelsInStorage();
+
+            // Act & Assert
+            Assert.ThrowsAsync<Exception>(() => this.target.GetListAsync());
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItReturnsAListOfCustomDeviceModels()
+        {
+            // Arrange
+            const string UPDATED_ETAG = "etag";
+            this.SetupAListOfDeviceModelsInStorage(UPDATED_ETAG);
+
+            // Act
+            var result = this.target.GetListAsync().Result;
+
+            // Assert
+            foreach (var model in result)
+            {
+                Assert.Equal(DeviceModel.DeviceModelType.CustomModel.ToString(), model.Type);
+                Assert.Equal(UPDATED_ETAG, model.ETag);
+            }
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItThrowsExceptionWhenGetDeviceModelByIdFailed()
+        {
+            // Act & Assert
+            Assert.ThrowsAsync<Exception>(() => this.target.GetAsync(String.Empty));
+        }
+
+        private void SetupAListOfDeviceModelsInStorage(string etag)
+        {
+            var deviceModel = new DeviceModel { Id = "id", ETag = etag };
+            var list = new ValueListApiModel();
+            var value = new ValueApiModel
+            {
+                Key = "key",
+                Data = JsonConvert.SerializeObject(deviceModel),
+                ETag = deviceModel.ETag
+            };
+            list.Items.Add(value);
+
+            this.storage
+                .Setup(x => x.GetAllAsync(It.IsAny<string>()))
+                .ReturnsAsync(list);
+        }
+
+        private void SetupAListOfInvalidDeviceModelsInStorage()
+        {
+            var obj = new { id = "id", eTag = "Etag" };
+            var list = new ValueListApiModel();
+            var value = new ValueApiModel
+            {
+                Key = "key",
+                Data = JsonConvert.SerializeObject(obj),
+                ETag = "etag"
+            };
+            list.Items.Add(value);
+
+            this.storage
+                .Setup(x => x.GetAllAsync(It.IsAny<string>()))
+                .ReturnsAsync(list);
+        }
+
+        private void SetupStorageToThrowException()
```

Comment:
for tests using this helper, I would find them more readable if this code was in the test (at the cost of duplication).  Also I would change this to throw some more interesting exception, and asserting the type matches

Reply:
Done.